### PR TITLE
fix(pdf-transcript): Don't error out when trying to process an image bigger than NATS max payload

### PR DESCRIPTION
**Pull request by KevLehman**
https://rocketchat.atlassian.net/browse/CORE-398

## Further comments

Fixed error handling for files bigger than NATS max allowed payload. This should prevent PDFs from erroring out when generating from rooms that contain heavy images. Also added projections to user/rooms db calls

### CODE REVIEW

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 1]
```diff
@@ -232,8 +232,20 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
...
+				} catch (e: unknown) {
+					this.log.error(`Failed to get file ${file._id}`, e);
+					// Push empty buffer so parser processes this as "unsupported file"
+					files.push({ name: file.name, buffer: null });
+
+					if ((e as Error).message === 'MAX_PAYLOAD_EXCEEDED') {
+						this.log.error(
+							`File is too big to be processed by NATS. See NATS config for allowing bigger messages to be sent between services`,
+						);
+					}
```
Comment (debdutdeb) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:247]: should we just increase or have some sort of file-split mechanism to work around it? like send 2MB-2MB-2MB -> 6MB file?

Comment (debdutdeb) [reply ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:247]: ps: can be done later, not a blocker for this pr imo.

Comment (KevLehman) [reply ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:247]: That would be cool, but yeah, for now, this is what we have.

Would be nice to send some streams over nats or maybe allowing the services to get a buffer directly with some package, idk.

(Or just chunked buffers, that works too)

File: .changeset/silly-clocks-return.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'@rocket.chat/omnichannel-services': patch
+'@rocket.chat/core-services': patch
+'@rocket.chat/meteor': patch
+---
+
+Fixed error handling for files bigger than NATS max allowed payload. This should prevent PDFs from erroring out when generating from rooms that contain heavy images. Also added projections to user/rooms db calls
```
Comment (debdutdeb) [inline .changeset/silly-clocks-return.md:7]:
```suggestion
Fixed error handling for files bigger than NATS max allowed payload. This should prevent PDFs from erroring out when generating from rooms that contain heavy images. 
```

**Review by debdutdeb — CHANGES_REQUESTED** [review]

File: apps/meteor/server/services/translation/service.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ export class TranslationService extends ServiceClassInternal implements ITransla
 	}
 
 	// Use translate when you want to translate to the user's language, or server's as a fallback
-	async translate(text: string, user: IUser): Promise<string> {
+	async translate(text: string, user: AtLeast<IUser, 'language'>): Promise<string> {
```
Comment (MarcosSpessatto) [inline apps/meteor/server/services/translation/service.ts:25]: Let's dismiss this unrelated change, also, according to the function implementation the language is not required.

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 4]
```diff
@@ -232,8 +232,20 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
...
-				files.push({ name: file.name, buffer: fileBuffer, extension: uploadedFile.extension });
+				try {
+					const fileBuffer = await uploadService.getFileBuffer({ file: uploadedFile });
+					files.push({ name: file.name, buffer: fileBuffer, extension: uploadedFile.extension });
+				} catch (e: unknown) {
+					this.log.error(`Failed to get file ${file._id}`, e);
+					// Push empty buffer so parser processes this as "unsupported file"
+					files.push({ name: file.name, buffer: null });
+
+					if ((e as Error).message === 'MAX_PAYLOAD_EXCEEDED') {
```
Comment (MarcosSpessatto) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:243]: Ideally, we should not tie this to the NATS error message, but since we don't have an abstraction for this, we could put a TODO comment to refactor this when it's possible. WDYT?

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 5]
```diff
@@ -275,7 +287,9 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
 		}
 		this.currentJobNumber++;
 		try {
-			const room = await LivechatRooms.findOneById(details.rid);
+			const room = await LivechatRooms.findOneById<Pick<IOmnichannelRoom, '_id' | 'v' | 'closedAt' | 'servedBy'>>(details.rid, {
+				projection: { _id: 1, v: 1, closedAt: 1, servedBy: 1 },
+			});
```
Comment (MarcosSpessatto) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:292]: It seems this is not related to the issue, if so, we can open a new PR for this improvement if necessary.

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 6]
```diff
@@ -346,11 +360,11 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
 
 	private async pdfFailed({ details, e }: { details: WorkDetailsWithSource; e: Error }): Promise<void> {
 		this.log.error(`Transcript for room ${details.rid} by user ${details.userId} - Failed: ${e.message}`);
-		const room = await LivechatRooms.findOneById(details.rid);
+		const room = await LivechatRooms.findOneById<Pick<IOmnichannelRoom, '_id'>>(details.rid, { projection: { _id: 1 } });
```
Comment (MarcosSpessatto) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:363]: Same as before

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 7]
```diff
@@ -346,11 +360,11 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
...
 		if (!room) {
 			return;
 		}
-		const user = await Users.findOneById(details.userId);
+		const user = await Users.findOneById<Pick<IUser, 'language' | '_id'>>(details.userId, { projection: { _id: 1, language: 1 } });
```
Comment (MarcosSpessatto) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:367]: Same

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 8]
```diff
@@ -369,7 +383,7 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
 
 	private async pdfComplete({ details, file }: { details: WorkDetailsWithSource; file: IUpload }): Promise<void> {
 		this.log.info(`Transcript for room ${details.rid} by user ${details.userId} - Complete`);
-		const user = await Users.findOneById(details.userId);
+		const user = await Users.findOneById<Pick<IUser, 'language' | '_id'>>(details.userId, { projection: { _id: 1, language: 1 } });
```
Comment (MarcosSpessatto) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:386]: Same

**Review by debdutdeb — DISMISSED** [review]

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 9]
```diff
@@ -232,8 +232,20 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
...
 				}
 
-				const fileBuffer = await uploadService.getFileBuffer({ file: uploadedFile });
-				files.push({ name: file.name, buffer: fileBuffer, extension: uploadedFile.extension });
+				try {
+					const fileBuffer = await uploadService.getFileBuffer({ file: uploadedFile });
+					files.push({ name: file.name, buffer: fileBuffer, extension: uploadedFile.extension });
+				} catch (e: unknown) {
+					this.log.error(`Failed to get file ${file._id}`, e);
+					// Push empty buffer so parser processes this as "unsupported file"
```
Comment (MarcosSpessatto) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:240]: I didn't find where this is being handled, can you show me where this is being handled as an unsupported file?

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts (hunk 9 above)
Comment (KevLehman) [reply ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:240]: That happens inside `pdf-worker`. When a `null` buffer is provided, it'll render an "Unsupported file" box on the transcript.

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts [hunk 10]
```diff
@@ -346,7 +359,7 @@ export class OmnichannelTranscript extends ServiceClass implements IOmnichannelT
 
 	private async pdfFailed({ details, e }: { details: WorkDetailsWithSource; e: Error }): Promise<void> {
 		this.log.error(`Transcript for room ${details.rid} by user ${details.userId} - Failed: ${e.message}`);
-		const room = await LivechatRooms.findOneById(details.rid);
+		const room = await LivechatRooms.findOneById<Pick<IOmnichannelRoom, '_id'>>(details.rid, { projection: { _id: 1 } });
```
Comment (KevLehman) [inline ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:362]:
```suggestion
		const room = await LivechatRooms.findOneById(details.rid);
```

File: ee/packages/omnichannel-services/src/OmnichannelTranscript.ts (hunk 1 above)
Comment (sampaiodiego) [reply ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:248]: moleculer supports Streams so in theory we could use them to transfer bigger files.. so after receiving the whole stream we could create a buffer from it again.. even better, we could change our APIs to support streams as well

Comment (KevLehman) [reply ee/packages/omnichannel-services/src/OmnichannelTranscript.ts:248]: That would be pretty cool! I can spend some time investigating that, since our current way of sending the whole buffer seems a bit overkill. We just do that with images but yeah.

**Review by MarcosSpessatto — APPROVED** [review]
Unit tests will be added in a subsequent PR.
